antrea-io/antrea
# WIP: Add yaml manifest to cleanup all Antrea resources

This is still very much a work-in-progress, I'm opening this PR to
gather feedback on the approach.

For someone deleting Antrea, the steps would be as follows:
 * `kubectl delete -f <path to antrea.yml>`
 * `kubectl apply -f <path to antrea-cleanup.yml>`
 * check that job has completed with `kubectl -n kube-system get jobs`
 * `kubectl delete -f <path to antrea-cleanup.yml>`

The cleanup manifest creates a DaemonSet that will perform the necessary
deletion tasks on each Node. After the tasks have been completed, the
"status" is reported to the cleanup controller through a custom
resource. Once the controller has received enough statuses (or after a
timeout of 1 minutes) the controller job completes and the user can
delete the cleanup manifest.

Known remaining items:
 * place cleanup binaries (antrea-cleanup-agent and
 antrea-cleanup-controller) in a separate docker image to avoid
 increasing the size of the main Antrea docker image
 * generate manifest with kustomize?
 * find a way to test this as part of CI?
 * update documentation
 * additional cleanup tasks: as of now we only take care of deleting the
 OVS bridge
 * place cleanup CRD in non-default namespace
 * use kubebuilder instead of the code generator directly (related to
 #16); we probably want to punt this task to a future PR.

See #181
